### ISSUE TITLE
Flow 0.53 prop types

### ISSUE THREAD

**Issue by pascalduez** [yannickcr/eslint-plugin-react#1376]
Hi,

with the latest [`Flow@0.53`](https://github.com/facebook/flow/blob/v0.53/Changelog.md) the Type detection for prop types is broken.

Before

```js
import React from 'react';

type Props = {
  foo: string,
};

class Bar extends React.Component {
  props: Props;
}
```

After

```js
import * as React from 'react';

type Props = {
  foo: string,
};

class Bar extends React.Component<Props> {}
```

```
'foo' is missing in props validation (react/prop-type)
```

**Comment by jseminck** [comment]
We currently have test cases like this. Are these still valid with Flow 0.53?

```js
        type Person = {
          firstname: string
        };
        class Hello extends React.Component<void, { person: Person }, void> {
          render () {
            return <div>Hello {this.props.person.lastname}</div>;
          }
        }
```

**Comment by ljharb** [comment]
Why `import * as React from 'react'`? There's no need to do that; `import React from 'react';` is correct.

**Comment by EvHaus** [comment]
@ljharb I was wondering about that too. However, Flow's official documentation says this change to the `import` is required and their codemods make this change as well (see https://github.com/facebook/flow/blob/a4df443a4a4fb5f62b86f03ee8a12a2b1432296a/website/en/docs/react/types.md).

Not sure if we can get someone from the Flow team to chime in, but it seems totally unnecessary to me as well (and potentially troublesome when using Webpack's Tree Shaking).

**Comment by ljharb** [comment]
That's horrific; it's one of the least usable parts of TypeScript.

**Comment by ljharb** [comment]
and yes, it also makes tree-shaking *impossible* unless you can statically determine every key accessed on a ModuleRecord.

**Comment by pascalduez** [comment]
Yep, this migh not be needed for the dummy example provided above, but will be as soon as you need one of the exposed types.
And yes the codemod convert all those imports anyway.

https://github.com/facebook/flow/blob/v0.53/Changelog.md#0530

**Comment by benjaminRomano** [comment]
I assume they are using the `import * as ...` syntax because they want people to use their types prefixed with `React`.

For example,
```js
import * as React from 'react';

let element: React.Node = ...;
```

is equivalent to 

```js
import React from 'react';
import type {Node} from 'react';

let element: Node = ...;
```
However, the first example is "cleaner".

**Comment by ljharb** [comment]
If the types are in the `react` package, then they should be importable directly. If they're not normal named imports, `import *` shouldn't be able to bring them in, per spec.

**Comment by tgallacher** [comment]
This is also a problem I have come across after upgrading to `Flow@0.53` with `eslint-plugin-react@7.2.1. 

I think Flow wants you explicitly use Reacts [utility types](https://flow.org/en/docs/react/types/), and `types` are not automatically attached to the `default` export as it isn't always guaranteed to be an object -- could be a number, for example -- and so you need to do a full import under a namespace. If you only need a single type, and want to take advantage of tree-shaking you could combine them in one line like: 

`import React, { type Node } form 'react'`

Note, theres a brief outline of this towards the end of the Components section on [Flow docs](https://flow.org/en/docs/react/components/#toc-class-components)

**Comment by ljharb** [comment]
The outline makes sense; in that case it's a poor choice for a codemod unless it's super smart about when it uses *.

**Comment by tgallacher** [comment]
Temp fix appears to be to use both the new required generics approach enforced by Flow@0.53, and the (now) legacy approach. For example: 

```javascript
import React, { type Element } from 'react'

type Props = {
   exA: number
};

type State = {
   exB: string
};


// 
// Latest version of Flow wants you to use generics
// but current version of Eslint plugin warns against react/prop-types
class Foo extends React.Component<Props, State>{
    props: Props; // This will keep "eslint-plugin-react" happy without interfering with Flow new generics req.

    state = {
        exB: 'initial val'
    };

    render(): Element<'div'>{
        const { exA } = this.props;
        const { exB } = this.state;

        return (
            <div>
                <ul>
                    <li>Props val: { exA }</li>
                    <li>(initial) State val: { exB }</li>
                </ul>
            </div>
        );
    }
}
```

**Comment by thejameskyle** [comment]
To summarize with the exact reasoning:

```js
import React from 'react'; // `React` is an object
import * as React from 'react'; // `React` is a namespace
```

Types can exist in namespaces, but not in objects.

You can also import specific bindings for both JavaScript values and Flow types:

```js
import { Component } from 'react';
import type { Node } from 'react';
import { type Node } from 'react';
```

And combine them if you want to import both the object and a type:

```js
import React, { type Node } from 'react';
```

**Comment by thejameskyle** [comment]
Because the type parameter signature changes in Flow 0.53 from:

```js
React.Component<DefaultProps, Props, State>
```

to:

```js
React.Component<Props>
// or 
React.Component<Props, State>
```

This ESlint plugin could be updated to pull props from:

```js
React.Component { props: Props; }
React.Component<X, Props, X> {}
React.Component<Props, X> {}
React.Component<Props> {}
```

And it will support all versions of Flow

**Comment by jseminck** [comment]
@thejameskyle in flow >0.53, was this also valid syntax: `React.Component<X, Props> {}` ? (2 arguments, where Props is the second argument)

If so, then we cannot distinguish properly between the old and the new syntax. If we expect `Props` to be the first or the second `TypedArgument`, then we'll be considering `State` as `Props` for those users who use `React.Component<Props, State> {}`, right?

Ofcourse we could do it so that if there are 3 arguments we check the second argument, and if there are 2 or 1 then we check the first. I just assumed that people may already have code like `React.Component<X, Props> {}` with the older Flow and then that would break.

The current idea we discussed in the PR is to introduce a `flowVersion` configuration, similar to how we already have a `version` configuration for React.

**Comment by thejameskyle** [comment]
No, in previous versions of Flow, you could only have no args, or three args:

```js
// Flow <0.53
class extends React.Component {}
class extends React.Component<DefaultProps, Props, State> {}

class extends React.Component { props: Props; }
class extends React.Component<DefaultProps, Props, State> { props: Props; }

// Flow >=0.53
class extends React.Component {}
class extends React.Component<Props> {}
class extends React.Component<Props, State> {}

class extends React.Component { props: Props; }
class extends React.Component<Props> { props: Props; }
class extends React.Component<Props, State> { props: Props; }
```

**Comment by jseminck** [comment]
Great. I think we can use that approach then! 👍 

3 arguments: check the second argument.
2 or 1 argument: check the first argument.

Thanks for your input.

**Comment by oliviertassinari** [comment]
> Why import * as React from 'react'? 

One important limitation of this change is around `React.propTypes` and `React.createClass` that are accessed. With react v15.x, it's raising two warnings:
![capture d ecran 2017-08-23 a 20 47 22](https://user-images.githubusercontent.com/3165635/29690202-52e2c210-8926-11e7-824c-7fe7b4abaf2e.png)

Explicitly importing the types is more appropriate until React v16 is released as stable.

**Comment by ljharb** [comment]
@oliviertassinari this is because you're supposed to use the `prop-types` and `create-react-class` packages as of v15.5, which work all the way back to 0.13 - so `React.PropTypes` isn't ever appropriate anymore anyways.

**Comment by oliviertassinari** [comment]
@ljharb It's a different topic. This change introduces the warnings with the latest version of react:
```diff
-import React from 'react';
+import * as React from 'react';
```

**Comment by ljharb** [comment]
ohhh i see what you mean, gotcha. that seems like a reason to use `import React from 'react'` instead anyways.

**Comment by jseminck** [comment]
This is now in master. It would be great if someone could try on their codebase(s) and confirm everything works!

**Comment by johnhaley81** [comment]
@jseminck I pulled in https://github.com/yannickcr/eslint-plugin-react#e2f6460a63bd0cd0dca0c8308e35ff831d5161ed but I'm still not getting prop validation detection with flow types. I also added:

```
"settings": {
    "react": {
      "flowVersion": "0.53"
    }
  }
```

to my `.eslintrc.js` file. Is there something else I'm missing?

**Comment by jseminck** [comment]
Thanks for testing it, @johnhaley81. Can you confirm you are using the master release? I did this to get it:

```js
rm -rf node_modules/eslint-plugin-react
yarn add https://github.com/yannickcr/eslint-plugin-react -D
```

One way to test it is to execute this command: `cat node_modules/eslint-plugin-react/lib/rules/prop-types.js | grep versionUtil` and it should return:

```js
const versionUtil = require('../util/version');
propsParameterPosition = versionUtil.testFlowVersion(context, '0.53.0') ? 0 : 1;
```

You shouldn't have to add the `flowVersion` to the config yet, though it doesn't hurt for the future. It should definitely work without.

Here is my example which does seem to work. `bar` is recognised, but `foo` is not. If you still have unexpected output then please provide a snippet and I'll look into it. I realise now that it is time for me to start using Flow in some side projects. 😄 

<img width="692" alt="screen shot 2017-08-26 at 23 57 57" src="https://user-images.githubusercontent.com/9586897/29745066-da1c648a-8aba-11e7-89c5-8bc884ac420d.png">

**Comment by johnhaley81** [comment]
I Confirmed that I'm on master. Here's a screenshot of what I'm seeing:

![image](https://user-images.githubusercontent.com/867780/29745762-dbc4f41a-8a78-11e7-95af-60a6201294ca.png)

**Comment by johnhaley81** [comment]
Oh and I would definitely recommend Flow. I've been using it for a few things and it's been great.

**Comment by ljharb** [comment]
@johnhaley81 i'm confused; what's `<P: { open?: boolean }>` doing there exactly? I'd expect something like `type P = { open?: boolean };` higher up in the file.

**Comment by jseminck** [comment]
Did this example work before, but using `<void, Props, void>` instead of `<Props>` ? Or is this syntax some new feature in the latest Flow version?

**Comment by thejameskyle** [comment]
@ljharb In Flow generics can have a "bound":

```js
class Foo<T: boolean> {
  constructor(val: T) { /* ... */ }
}

let a = new Foo(true); // works
let b = new Foo(42); // error
```

Additionally, generics can be used in nested scopes:

```js
function method<T>() {
  return function(val: T) { /* ... */ }
}
```

So when you put those together you have:

```js
function higherOrderComponent<Props: { ... }>() {
  return class extends React.Component<Props> {
    // ...
  }
}
```

It looks like this plugin might not be looking up the type of `Props` from scope well enough. 

I have a few Babel util packages to help with this:

- https://github.com/babel-utils/babel-identifiers
- https://github.com/babel-utils/babel-flow-identifiers
- https://github.com/babel-utils/babel-flow-scope

```js
import {getIdentifierKind} from 'babel-identifiers';
import {isFlowIdentifier} from 'babel-flow-identifiers';
import {findFlowBinding} from 'babel-flow-scope';

if (getIdentifierKind(path) === 'reference') {
  let name = path.node.name;
  let binding;

  if (isFlowIdentifier(path)) {
    binding = findFlowBinding(path, name);
  } else {
    binding = path.scope.getBinding(name);
  }

  // ...
}
```

Hopefully these can help.

**Comment by jseminck** [comment]
Shouldn't this be raised as a separate issue?

**Comment by ljharb** [comment]
This issue is a generic one for 0.53 prop types support; are bound generics new in 0.53?

**Comment by jseminck** [comment]
Not sure if it is. I guess it might be. Anyway, I had a quick look at this and identified at least one issue. We are only visiting `ClassDeclaration` node types, but when using HOC pattern we return a `ClassExpression`:

<img width="1360" alt="screen shot 2017-08-27 at 19 59 35" src="https://user-images.githubusercontent.com/9586897/29752204-6b3884a6-8b62-11e7-8ce7-e249563dec70.png">

Adding support for this type makes these two cases valid (previous they would fail):

```js
        type Props = { foo: string }
        function higherOrderComponent<Props>() {
          return class extends React.Component<Props> {
            render() {
              return <div>{this.props.foo}</div>
            }
          }
        }
```

And

```js
function higherOrderComponent<P: { foo: string }>() {
          return class extends React.Component<P> {
            render() {
              return <div>{this.props.foo}</div>
            }
          }
        }
```

Finally, also added a test case for @johnhaley81 's issue and it seems to pass after these changes:

```js
        const withOverlayState = <P: {foo: string}>(WrappedComponent: ComponentType<P>): CpmponentType<P> => (
          class extends React.Component<P> {
            constructor(props) {
              super(props);
              this.state = {foo: props.foo}
            }
            render() {
              return <div>Hello World</div>
            }
          }
        )
```

PR coming soon 🎉

**Comment by jasonbarry** [comment]
This has landed in master, yes? Can a patch release come out soon? Right now any project using latest flow and latest eslint-plugin-react won't pass linting.

**Comment by jseminck** [comment]
Partially landed in master, yes. There is still one PR that solves another case.

**Comment by ardok** [comment]
Can't wait to see this closed!

**Comment by silvenon** [comment]
Let's be patient and buy @jseminck a beer afterwards. 😄

**Comment by alexreardon** [comment]
Thanks for working on this everyone - it now works great!
